src/libcrypto/poly_math.c

```diff
@@ -1,39 +1,254 @@
 #include "include_asm.h"
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", null);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", double_null);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", double_add);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", copy);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", double_copy);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", single_to_double);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", double_to_single);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", multiply_shift);
+#include "libcrypto/bb_nn.h"
+#include "bbtypes.h"
+#include "macros.h"
+
+void null(field_2n* a) {
+    element* eptr = a->e;
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(a->e); i++) {
+        *eptr++ = 0;
+    }
+}
+
+void double_null(field_double* a) {
+    element* eptr = a->e;
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(a->e); i++) {
+        *eptr++ = 0;
+    }
+}
+
+void double_add(field_double* a, field_double* b, field_double* c) {
+    element* aptr = a->e;
+    element* bptr = b->e;
+    element* cptr = c->e;
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(a->e); i++) {
+        int tempA = *aptr++;
+        int tempB = *bptr++;
```

Comment:
```suggestion
        unsigned long tempA = *aptr++;
        unsigned long tempB = *bptr++;
```
Does this still match? `element` is an `unsigned long` so it would be more correct